File: python_src/399.除法求值.py
#
# @lc app=leetcode.cn id=399 lang=python3
#
# [399] 除法求值
#
from typing import *
from collections import *

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        eqt = defaultdict(list)
        ret = []
        for eq, v in zip(equations, values):
            eqt[eq[0]].append((eq[1], v))
            eqt[eq[1]].append((eq[0], 1./v))

        self.eqt = eqt
        self.access = set()
        
        for a, b in queries:
            self.access.clear()
            if a not in eqt or b not in eqt:
                ret.append(-1.)
                continue
            ret.append(self.dfs(a, b))
        return ret

    def dfs(self, v: str, target: str) -> float:
        if v in self.access:
            return -1.
        self.access.add(v)
        for deno, frac in self.eqt[v]:
            if deno == target:
                return frac
            r = self.dfs(deno, target)
            if r != -1.:
                return r * frac
        return -1.
        
# @lc code=end

s = Solution()
print(s.calcEquation(
equations = [["a","b"],["b","c"]], values = [2.0,3.0], queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]
))
print(s.calcEquation(
equations = [["a","b"],["b","c"],["bc","cd"]], values = [1.5,2.5,5.0], queries = [["a","c"],["c","b"],["bc","cd"],["cd","bc"]]
))
print(s.calcEquation(
equations = [["a","b"]], values = [0.5], queries = [["a","b"],["b","a"],["a","c"],["x","y"]]
))

File: cpp_src/399.除法求值.cpp
/*
 * @lc app=leetcode.cn id=399 lang=cpp
 *
 * [399] 除法求值
 */
#include "utils.hpp"

using namespace std;

// @lc code=start
class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, unordered_map<string, double>> graph;
        vector<double> ans;
        for (size_t i = 0; i < values.size(); i++) {
            graph[equations[i][0]][equations[i][1]] = values[i];
            graph[equations[i][1]][equations[i][0]] = 1. / values[i];
        }
        for (auto& query : queries) {
            double res = -1.;
            if (graph.contains(query[0]) && graph.contains(query[1])) {
                res = query[0] == query[1] ? 1. : bfs(graph, query[0], query[1]);
            }
            ans.emplace_back(res);
        }
        return ans;
    }

    double bfs(unordered_map<string, unordered_map<string, double>>& graph,
        const string& src, const string& target) {
        if (graph[src].contains(target))
            return graph[src][target];

        queue<string> q;
        unordered_set<string> visited;
        q.emplace(src);
        visited.emplace(src);
        while (q.size()) {
            auto& node = q.front();
            for (auto& t : graph[node]) {
                if (visited.find(t.first) == visited.end()) {
                    visited.emplace(t.first);
                    q.emplace(t.first);
                    auto val = t.second * graph[src][node];
                    graph[src].emplace(t.first, val);
                    graph[t.first].emplace(src, 1. / val);
                }
            }
            q.pop();
        }
        return graph[src].contains(target) ? graph[src][target] : -1.;
    }
};
// @lc code=end

void test() {
    Solution s;
    vector<vector<string>> equations;
    vector<double> values;
    vector<vector<string>> queries;

    equations = { {"a", "b"}, {"b", "c"} };
    values = { 2.0, 3.0 };
    queries = { {"a", "c"}, {"b", "a"}, {"a", "e"}, {"a", "a"}, {"x", "x"} };
    cout << s.calcEquation(equations, values, queries) << endl;

    equations = { {"a", "b"}, {"b", "c"}, {"bc", "cd"} };
    values = { 1.5, 2.5, 5.0 };
    queries = { {"a", "c"}, {"c", "b"}, {"bc", "cd"}, {"cd", "bc"} };
    cout << s.calcEquation(equations, values, queries) << endl;

    equations = { {"a", "b"} };
    values = { 0.5 };
    queries = { {"a", "b"}, {"b", "a"}, {"a", "c"}, {"x", "y"} };
    cout << s.calcEquation(equations, values, queries) << endl;
}
